### ISSUE TITLE
Design without repository pattern

### ISSUE THREAD

**Issue by fabercs** [dotnet-architecture/eShopOnWeb#350]
Is there a way to implement this design without repository pattern, by calling dbContext directly from controllers and services?

From controllers side (web) it is ok, however, from services(Core) it causes circular reference, any thoughts how to prevent it?

**Comment by fabercs** [comment]
Well, i have done this with the approach below;

```
public class BaseService
{
     protected readonly DbContext DbContext;
     public BaseService(IServiceProvider serviceProvider)
     {
         var factory = serviceProvider.GetService<IDesignTimeDbContextFactory<DbContext>>();
         DbContext ??= factory.CreateDbContext(new string[] { });
     }
 }
```

```
public class TenantService : BaseService, ITenantService
{
    public TenantService(IServiceProvider provider)
        : base(provider)
    {
    }
    public async Task<Guid> CreateTenant(Tenant tenant)
    {
        var tenants = DbContext.Set<Tenant>();
        tenants.Add(tenant);

        await DbContext.SaveChangesAsync();
        return tenant.Id;
    }

    public Task<List<Tenant>> GetAll()
    {
        var tenants = DbContext.Set<Tenant>();
        return tenants.ToListAsync();
    }
}
```
And this is the context factory method;

```
public class HostDbContextFactory : IDesignTimeDbContextFactory<HostDbContext>
{
    public HostDbContext CreateDbContext(string[] args)
    {
        var envName = Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT");

        var configuration = new ConfigurationBuilder()
           .SetBasePath(Path.Combine(Directory.GetCurrentDirectory()))
           .AddJsonFile("appsettings.json", optional: false)
           .AddJsonFile($"appsettings.{envName}.json", optional: false)
           .Build();

        var builder = new DbContextOptionsBuilder<HostDbContext>();

        var connString = configuration.GetConnectionString("HostConnectionString");
        builder.UseSqlServer(connString);

        return new HostDbContext(builder.Options);
    }
}
```

Do you think is this a good design/approach?

**Comment by efleming18** [comment]
I don't think I really see a problem with what you have above - I will think on it a bit more but wanted to get you a reply.

One thing I would note, is that if you did plan to use the DB context in a controller, I would recommend using something like MediatR (which we have an example of in this project) so that the controller itself doesn't actually have a dependency on the DB context. It removes the dependencies / logic from the controller itself.

Just for curiosity, why are you _not_ wanting to use the repository pattern?

**Comment by fabercs** [comment]
Hi Eric, thank you for answer. 

I have seen Meditor pattern earlier, but never had any effort to grasp and use it in any project. As I understand, it creates another layer of abstraction to reduce the complexity of controller's dependencies. Well, I think as a mid project of mine, this implementation would increase the effort and complexity instead.

For repository pattern, as you all know, it is a debate for a long time whether this is needed or not where that `DbContext` already implements Repository pattern and UoW.
[https://gunnarpeipman.com/ef-core-repository-unit-of-work/](https://gunnarpeipman.com/ef-core-repository-unit-of-work/)
I couldn't find it now , but even in microsoft docs it is recommended that you do not need to wrap dbcontext with any implementation of Repo or UoW. 
So, I just wanted to go on without repository pattern for my new fresh project to see what's gonna happen :)

**Comment by ardalis** [comment]
Yeah, those articles are misleading, IMO. There are a bunch of benefits and virtually no downsides to programming to an abstraction versus an implementation. Yes, EF implements repository and UoW, but the point of those patterns isn't in the implementation but in the abstraction. And ideally, in you owning your abstraction, not choosing to depend on a vendor's.

Let me know how it works out for you, and you certainly *can* build software while tightly coupling to dependencies, I've just been bitten by enough of it to have strong opinions against it for anything that needs to be maintained for more than a little while. You'll find more of my opinions on this here if you're interested:

- https://www.weeklydevtips.com/episodes/024
- https://www.weeklydevtips.com/episodes/025
- https://www.weeklydevtips.com/episodes/026

**Comment by efleming18** [comment]
Hey @fabercs - any other questions here before I close this issue out?

**Comment by fabercs** [comment]
Hey @efleming18, until now, I have been into this, I drown as I dive into the deep, lots of opinions...
So, I decided to use a generic repository just because of the plug on/off idea of the persistence vendor, thanks to the @ardalis 's podcasts. Thank you for your guidance.

However, this time I started to research and think about the mediator pattern (MediatR library). I also came up to another popular asp.net core clean architecture starting point project on github. It was orchestrating its entire service layer with MediatR. There was not so much coverage on real-world problems, so I couldn't make a keen decision. Although, it is a very simple pattern, I can not understand how this pattern fits instead of a repo-service layer, besides, I can already see that I will have a few hundreds of command and query objects. Does it really worth it to replace service layer to a mediatR pattern?
